src/node/chainstate.cpp

```diff
@@ -95,11 +95,12 @@ static ChainstateLoadResult CompleteChainstateInitialization(
     assert(chainman.m_total_coinstip_cache > 0);
     assert(chainman.m_total_coinsdb_cache > 0);
 
-    // Conservative value which is arbitrarily chosen, as it will ultimately be changed
-    // by a call to `chainman.MaybeRebalanceCaches()`. We just need to make sure
-    // that the sum of the two caches (40%) does not exceed the allowable amount
-    // during this temporary initialization state.
-    double init_cache_fraction = 0.2;
+    // If running with multiple chainstates, limit the cache sizes with a
+    // discount factor. If discounted the actual cache size will be
+    // recalculated by `chainman.MaybeRebalanceCaches()`. The discount factor
+    // is conservatively chosen such that the sum of the caches does not exceed
+    // the allowable amount during this temporary initialization state.
+    double init_cache_fraction = chainman.GetAll().size() > 1 ? 0.2 : 1.0;
```

Comment:
Is this the same as
```suggestion
    auto init_cache_fraction = chainman.IsSnapshotActive() ? 0.2 : 1.0;
```
?

Reply:
I think in practice it is, since we call this after `DetectSnapshotChainstate()`, which in turn calls `ActivateExistingSnapshot` if there is a snapshot chainstate, but checking if there is more than one chainstate seems more defensive to me, since it relies on fewer assumptions.

Reply:
fyi I also hope we can improve how to do this in ryanofsky's #30214.